calico_containers/pycalico/ipam.py

```diff
@@ -832,3 +1034,116 @@ def get_assignment_attributes(self, address):
+
+        :param cidr: The CIDR covering the blocks to be released.  Raises a
+        InvalidBlockSizeError if the CIDR is smaller than the minimum allowable
+        block size.
+        :param host: (optional) The host ID to use for affinity in assigning IP
+        addresses.  Defaults to the hostname returned by get_hostname().
+
+        :return: A tuple of:
+                 ([IPNetwork<blocks claimed>],
+                  [IPNetwork<blocks that were claimed by another host>])
+        """
+        assert isinstance(cidr, IPNetwork)
+        if not validate_block_size(cidr):
+            _log.info("Requested CIDR %s is too small", cidr)
+            raise InvalidBlockSizeError("The requested CIDR is smaller than "
+                                        "the minimum block size.")
+
+        if not host:
+            host = get_hostname()
+
+        if self._validate_cidr_in_pools(cidr):
+            _log.info("Requested CIDR %s is not is a configured pool", cidr)
+            raise PoolNotFound("Requested CIDR is not in a configured IP "
+                               "Pool.")
+
+        claimed = []
+        unclaimed = []
+
+        # Get the IPAM configuration.  We need this when claiming block
+        # affinities.
+        ipam_config = self.get_ipam_config()
+
+        for block_cidr in cidr.subnet(BLOCK_PREFIXLEN[cidr.version]):
+            try:
+                self._claim_block_affinity(host, block_cidr, ipam_config)
+            except HostAffinityClaimedError:
+                unclaimed.append(cidr)
+                break
+            else:
+                claimed.append(cidr)
+
+        return claimed, unclaimed
+
+    @handle_errors
+    def release_affinity(self, cidr, host=None):
+        """
+        :param cidr: The CIDR covering the blocks to be released.  Raises a
+        InvalidBlockSizeError if the CIDR is smaller than the minimum allowable
+        block size.
+        :param host: (optional) The host ID to compare against the affinity of
+        each block that is being released.
+
+        :return: A tuple of:
+                 ([IPNetwork<blocks released>],
+                  [IPNetwork<blocks that were not claimed>],
+                  [IPNetwork<blocks that were claimed by another host>])
+        """
+        assert isinstance(cidr, IPNetwork)
+        if not validate_block_size(cidr):
+            _log.info("Requested CIDR %s is too small", cidr)
+            raise InvalidBlockSizeError("The requested CIDR is smaller than "
+                                        "the minimum block size.")
+        if not host:
+            host = get_hostname()
+
+        released = []
+        not_claimed = []
+        claimed_by_other = []
+
+        for block_cidr in cidr.subnet(BLOCK_PREFIXLEN[cidr.version]):
+            try:
+                self._release_block_affinity(host, block_cidr)
+            except HostAffinityClaimedByOtherError:
+                claimed_by_other.append(cidr)
+                break
```

Comment:
What is the reasoning for wanting to `break` here (and below), rather than continue and release as many of the blocks that we own as we can?